backend/store/etcd/event_store.go

```diff
@@ -203,34 +203,58 @@ func (s *Store) GetEventByEntityCheck(ctx context.Context, entityName, checkName
 }
 
 // UpdateEvent updates an event.
-func (s *Store) UpdateEvent(ctx context.Context, event *corev2.Event) error {
+func (s *Store) UpdateEvent(ctx context.Context, event *corev2.Event) (*corev2.Event, *corev2.Event, error) {
 	if event == nil || event.Check == nil {
-		return errors.New("event has no check")
+		return nil, nil, errors.New("event has no check")
 	}
 
 	if err := event.Check.Validate(); err != nil {
-		return err
+		return nil, nil, err
 	}
 
 	if err := event.Entity.Validate(); err != nil {
-		return err
+		return nil, nil, err
+	}
+
+	ctx = store.NamespaceContext(ctx, event.Entity.Namespace)
+
+	prevEvent, err := s.GetEventByEntityCheck(
+		ctx, event.Entity.Name, event.Check.Name,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Maintain check history.
+	if prevEvent != nil {
+		if !prevEvent.HasCheck() {
+			return nil, nil, errors.New("invalid previous event")
+		}
+
+		event.Check.MergeWith(prevEvent.Check)
 	}
 
+	updateOccurrences(event.Check)
```

Comment:
Would it make sense to have this business logic in the `api/core/v2` package too? Or it's only relevant for etcd?

Reply:
It's something that's going to be different for each store unfortunately.